### PR TITLE
HDDS-7016. EC: Implement the Over replication Processor

### PR DESCRIPTION
**Pull request by umamaheswararao**
## What changes were proposed in this pull request?

Implemented the processing logic of OverReplicated containers of EC. The logic is very similar to UnderReplicated containers processing. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7016

## How was this patch tested?

Added test

### CODE REVIEW

**Comment by umamaheswararao** [comment]
Thank you @JacksonYao287 for the review!
I think the thought was that, RM is always do full loop of rechecking. So, if we maintain the same queue and just add we get log of duplicates would appear. The idea here is, when RM does re-loop, that should get the latest full set, and that replaces the queue itself. Just for adding the failed things back of the queue we are using these queues and for under replication we are doing prioritizations. 
They both needs to run dedicatedly. Not sure having thread pool make much difference.

As per @sodonnel both of these threads needs the improvements with respective to DN load considerations. That is the next work we have there.

**Comment by umamaheswararao** [comment]
@JacksonYao287 Let's discuss the thread pool thought, if that's helpful here, I will take up in the followup JIRA. Thank you

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationManager.java [hunk 1]
```diff
@@ -295,6 +302,22 @@ public synchronized void processAll() {
     }
   }
 
+  /**
+   * Retrieve the new highest priority container to be replicated from the
+   * under replicated queue.
+   * @return The new underReplicated container to be processed, or null if the
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationManager.java:308]:
```suggestion
   * over-replicated queue.
   * @return The next over-replicated container to be processed, or null if the
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/TestECContainerRecovery.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.hadoop.hdds.scm.ScmConfigKeys.HDDS_SCM_WATCHER_TIMEOUT;
+
+/**
+ * Tests key output stream.
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/TestECContainerRecovery.java:62]: Stale comment?

**Review by adoroszlai — COMMENTED** [review]
Thanks @umamaheswararao for working on this.  LGTM, except some leftover comments.  We might want to look into reducing code duplication later.

**Comment by umamaheswararao** [comment]
Thanks @adoroszlai for the review. I have just addressed the comments. Thank you.

**Comment by sodonnel** [comment]
> 2 in ReplicationManager#processAll, we get two new instances of queue for both overReplicated and underReplicated tasks every time, and they are protected by a lock. there may be a case that the earlier generated task queue is replaced by a later generated task queue without completing all the tasks in the earlier one. for example, the interval of OverReplicatedProcessor is 10m, and the interval of RM is 2m. so in 10m, RM will produce 5 task queues for overreplicated containers, but OverReplicatedProcessor can only consume the latest one.

Not sure if an earlier commit had a different time, but the intention is that the under / over replication threads will run at a faster interval than the replication manager. Eg 30 seconds, as that is the DN heartbeart interval. The queues will be regenerated each time the RM thread runs, probably every 5 or 10 minutes.

As we are not using an event driven model where, for example, the dead node event causes the containers to be marked under replicated or a node registration causes them to be over replicated (this is how HDFS works), we decided to keep things simple and just re-create the queues on each iteration rather than trying to maintain a single queue.

**Review by sodonnel — APPROVED** [review]
LGTM

**Pull request merged**